openmc/model/model.py

```diff
@@ -207,17 +212,34 @@ def run(self, **kwargs):
         Returns
         -------
         uncertainties.UFloat
-            Combined estimator of k-effective from the statepoint
+            Combined estimator of k-effective from the last statepoint
+            (None if no statepoint was written)
 
         """
+
         self.export_to_xml()
 
-        openmc.run(**kwargs)
+        # Setting tstart here ensures we don't pick up any old statepoint
+        # files that might preexist in the output directory
+        tstart = time.time()
+        self.statepoint = None
 
-        n = self.settings.batches
-        if self.settings.statepoint is not None:
-            if 'batches' in self.settings.statepoint:
-                n = self.settings.statepoint['batches'][-1]
+        openmc.run(**kwargs)
 
-        with openmc.StatePoint('statepoint.{}.h5'.format(n)) as sp:
-            return sp.k_combined
+        # Get output directory and last statepoint written by this run
+        if self.settings.output and 'path' in self.settings.output:
+            output_dir = Path(self.settings.output['path'])
+        else:
+            output_dir = Path.cwd()
+        for sp in output_dir.glob('statepoint.*.h5'):
+            mtime = sp.stat().st_mtime
+            if mtime >= tstart:  # >= allows for poor clock resolution
+                tstart = mtime
+                self.statepoint = sp
+
+        # Open the last statepoint to get the final k-effective
+        keff = None
+        if self.statepoint:
+            with openmc.StatePoint(self.statepoint) as sp:
```

Comment:
Looks like this is needed to prevent issues on Python 3.5
```suggestion
            with openmc.StatePoint(str(self.statepoint)) as sp:
```